[Simplified Collect][Taxes] Tax editing and bulk actions

kosmydel opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/37794
$ https://github.com/Expensify/App/issues/37795
$ https://github.com/Expensify/App/issues/37796
$ https://github.com/Expensify/App/issues/37797

It also fixes:
$ https://github.com/Expensify/App/issues/38499
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go to Collect Workspace -> Taxes.
2. Click on tax rate to open the RHP modal.
3. Try to toggle the tax rate (enable/disable).
4. Try to change the tax name.
5. Try to change the tax value.
6. Validations:
   1. Edit to a duplicated name.
   2. Select an empty name.
   3. Select an empty value.
   4. Select a tax value of more than 100 percent.
   5. Try to change the value to "5.", save, verify the dot is removed.
   6. Try to change the value to "005", save, verify the leading zeros are removed.
7. Bulk actions:
   1. Select a few taxes and try to delete them.
   2. Select a few taxes and try to enable them.
   3. Select a few taxes and try to disable them.
8. Verify you cannot delete and disable the default tax.
9. Failing requests:
   1.  Turn on failing requests.
   2.  Try to edit something.
   3.  Verify the error appears (might take a while, as it retries a few times).
   4.  Verify the red brick is shown.
   5.  Dismiss the error.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

1. Turn on Offline mode.
2. Open edit tax rate modal.
3. Edit each field and verify that after editing, the field has half opacity.
4. Turn off Offline mode.
5. Verify the taxes are updated.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Tests from the above

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/104823336/edb3ad2c-cab1-491f-9e8e-5ae448e3f1f3



</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/104823336/36971d2b-b8d1-44ae-8840-02099009c50e

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/104823336/3a6de717-f8bf-4266-9ea2-ff7d353ef8c0


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/104823336/1cd57e4a-ba76-4069-a28a-9ccd4fce5eb5



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/104823336/0be4497d-99c4-4ab2-bfe7-885db0d30ed5



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/104823336/7fe4cc0d-a7cc-419b-9b9c-489ab24fd1f5



</details>
****

kosmydel committed:
add edit tax modal

kosmydel committed:
add enabling/disabling taxes

kosmydel committed:
add deleting tax rates

kosmydel committed:
update enable tax api call

kosmydel committed:
add NamePage and ValuePage

kosmydel committed:
Merge branch 'wave8/CreatingNewTaxes' into wave8/WorkspaceEditPage

kosmydel committed:
renaming tax names

kosmydel committed:
update backend queries

kosmydel committed:
refactor

kosmydel committed:
update to new backend

kosmydel committed:
add bulk actions

kosmydel committed:
add wrappers

kosmydel committed:
pending actions draft

kosmydel committed:
Merge branch 'wave8/CreatingNewTaxes' into wave8/WorkspaceEditPage

kosmydel committed:
fix error handling and offline feedback

DylanDylann commented:
@kosmydel It seems this PR will handle 4 issues:
- https://github.com/Expensify/App/issues/37794
- https://github.com/Expensify/App/issues/37795
- https://github.com/Expensify/App/issues/37796
- https://github.com/Expensify/App/issues/37797

DylanDylann commented:
@kosmydel It seems there are no blockers anymore. Please continue to process this PR

kosmydel commented:
I should be able to prepare this PR for review tomorrow.

kosmydel committed:
Merge branch 'main' into wave8/WorkspaceEditPage

kosmydel committed:
improvements

kosmydel committed:
improve validation

filip-solecki commented on src/ONYXKEYS.ts:
```diff
@@ -415,6 +415,8 @@ const ONYXKEYS = {
         POLICY_TAG_NAME_FORM_DRAFT: 'policyTagNameFormDraft',
         WORKSPACE_NEW_TAX_FORM: 'workspaceNewTaxForm',
         WORKSPACE_NEW_TAX_FORM_DRAFT: 'workspaceNewTaxFormDraft',
+        WORKSPACE_TAX_NAME_FORM: 'workspaceTaxNameForm',
```
Add drafts

filip-solecki commented on src/ROUTES.ts:
```diff
@@ -612,6 +612,18 @@ const ROUTES = {
         getRoute: (policyID: string) => `settings/workspaces/${policyID}/taxes/new` as const,
     },
+    WORKSPACE_TAXES_EDIT: {
+        route: 'workspace/:policyID/tax/:taxID',
```
adjust routes to new ideal-nav-v2

filip-solecki commented on src/languages/en.ts:
```diff
@@ -1871,6 +1871,12 @@ export default {
                 valuePercentageRange: 'Please enter a valid percentage between 0 and 100.',
                 genericFailureMessage: 'An error occurred while updating the tax rate, please try again.',
             },
+            deleteTaxConfirmation: 'Are you sure you want to delete this tax?',
```
missing es translations?

kosmydel replied:
Asked for translations [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1710843079162819).

filip-solecki commented on src/libs/PolicyUtils.ts:
```diff
@@ -276,6 +276,11 @@ function goBackFromInvalidPolicy() {
 
+/** Get a tax with given ID from policy */
+function getTaxByID(policy: OnyxEntry<Policy>, taxID: string): TaxRate | undefined {
+    return policy?.taxRates?.taxes?.[taxID ?? ''];
```
```suggestion
    return policy?.taxRates?.taxes?.[taxID];
```
Should be enough since taxID is always string

filip-solecki commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +144,300 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+        policyID,
+        taxCode: taxID,
+        taxAmount: Number(taxValue),
+    } as UpdatePolicyTaxValueParams;
```
why is this assertion needed?

filip-solecki replied:
Why do we need it?

filip-solecki replied:
We don't do it for other actions

kosmydel replied:
I mean it's not necessary, but it's a bit stricter type. If needed, I can remove it.

filip-solecki commented on src/pages/workspace/taxes/NamePage.tsx:
```diff
@@ -0,0 +1,100 @@
+                                value={name}
+                                maxLength={CONST.TAX_RATES.NAME_MAX_LENGTH}
+                                onChangeText={setName}
+                                multiline={false}
```
```suggestion

```

filip-solecki commented on src/pages/workspace/taxes/ValuePage.tsx:
```diff
@@ -0,0 +1,105 @@
+
+    const goBack = useCallback(() => Navigation.goBack(ROUTES.WORKSPACE_TAXES_EDIT.getRoute(policyID ?? '', taxID)), [policyID, taxID]);
+
+    // TODO: Extract it to a separate file, and use it also when creating a new tax
```
Maybe it can be done now? Or you prefer some follow-up?

kosmydel replied:
I've added it but forgot to push :D Check it now

filip-solecki commented on src/pages/workspace/taxes/WorkspaceEditTaxPage.tsx:
```diff
@@ -0,0 +1,155 @@
+        if (!policy?.id) {
+            return;
+        }
+        deletePolicyTaxes(policy?.id, [taxID]);
```
i think we agreed to use `route.params.policyID` in each place - check all files

kosmydel committed:
use policyID from route

kosmydel committed:
address review

filip-solecki commented on src/pages/workspace/taxes/WorkspaceEditTaxPage.tsx:
```diff
@@ -0,0 +1,158 @@
+        Navigation.goBack();
+    };
+
+    const threeDotsMenuItems = useMemo(() => {
+        const menuItems = [
+            {
+                icon: Expensicons.Trashcan,
+                text: translate('common.delete'),
+                onSelected: () => setIsDeleteModalVisible(true),
+            },
+        ] satisfies ThreeDotsMenuItem[];
+        return menuItems;
+    }, [translate]);
```
```suggestion
    const threeDotsMenuItems = useMemo(
        () => [
            {
                icon: Expensicons.Trashcan,
                text: translate('common.delete'),
                onSelected: () => setIsDeleteModalVisible(true),
            },
        ],
        [translate],
    );
```

filip-solecki commented on src/pages/workspace/taxes/WorkspaceEditTaxPage.tsx:
```diff
@@ -0,0 +1,158 @@
+                    style={styles.mb5}
+                >
+                    <View style={[styles.h100, styles.flex1, styles.justifyContentBetween]}>
+                        <View>
```
is it needed?

kosmydel committed:
new spanish translations

filip-solecki commented on src/pages/workspace/taxes/WorkspaceTaxesPage.tsx:
```diff
@@ -129,23 +134,87 @@ function WorkspaceTaxesPage({policy, route}: WorkspaceTaxesPageProps) {
     );
 
+    const deleteTaxes = useCallback(() => {
+        if (!route.params.policyID) {
```
maybe extract `policyID` from routes in the function declaration? We use it here many times

kosmydel committed:
new translations

kosmydel committed:
add param, address review

kosmydel committed:
extract policyID

kosmydel committed:
lint & ts

kosmydel committed:
address review

kosmydel committed:
improve types

kosmydel commented:
I think I have everything except for [this one](https://github.com/Expensify/App/issues/38499#issuecomment-2004931877). 

I'm going to record videos and work on the above later, to unblock the review process.

kosmydel committed:
fix routes

kosmydel committed:
fix icons,

kosmydel committed:
fix dropdown button

kosmydel committed:
fix android

luacmartins commented on src/ROUTES.ts:
```diff
@@ -612,6 +612,18 @@ const ROUTES = {
+        route: 'settings/workspaces/:policyID/tax/:taxID',
+        getRoute: (policyID: string, taxID: string) => `settings/workspaces/${policyID}/tax/${encodeURI(taxID)}` as const,
+    },
+    WORKSPACE_TAXES_NAME: {
+        route: 'settings/workspaces/:policyID/tax/:taxID/name',
+        getRoute: (policyID: string, taxID: string) => `settings/workspaces/${policyID}/tax/${encodeURI(taxID)}/name` as const,
+    },
+    WORKSPACE_TAXES_VALUE: {
+        route: 'settings/workspaces/:policyID/tax/:taxID/value',
+        getRoute: (policyID: string, taxID: string) => `settings/workspaces/${policyID}/tax/${encodeURI(taxID)}/value` as const,
+    },
```
NAB the url changes when navigating from Edit to Name/Value pages. We can address this as a follow up though

https://github.com/Expensify/App/assets/22219519/22f8124a-0903-439e-abe0-a3b4aec32e09

luacmartins commented on src/languages/en.ts:
```diff
@@ -1869,7 +1869,19 @@ export default {
                 taxRateAlreadyExists: 'This tax name is already in use.',
                 valuePercentageRange: 'Please enter a valid percentage between 0 and 100.',
-                genericFailureMessage: 'An error occurred while updating the tax rate, please try again.',
+                deleteFailureMessage: 'An error occurred while deleting the tax rate, please try again.',
+                updateFailureMessage: 'An error occurred while updating the tax rate, please try again.',
+                createFailureMessage: 'An error occurred while creating the tax rate, please try again.',
+            },
+            deleteTaxConfirmation: 'Are you sure you want to delete this tax?',
+            deleteMultipleTaxConfirmation: ({taxAmount}) => `Are you sure you want to delete ${taxAmount} taxes?`,
+            actions: {
+                delete: 'Delete rate',
+                deleteMultiple: 'Delete rates',
+                disable: 'Disable rate',
+                disableMultiple: 'Disable rates',
+                enable: 'Enable rate',
+                enableMultiple: 'Enable rates',
```
@Expensify/marketing could you please review this copy?

jamesdeanexpensify replied:
To confirm, these are just generic error messages where we don't really know exactly what happened? Or is there any more info we have where we can get more specific? Thanks!

luacmartins replied:
Yea, these are just generic for when we don't know what happened. Otherwise the server sends specific error messages for the cases we do.

kosmydel replied:
I think if needed we can be a bit more specific for `updateFailureMessage`. We can split it:
- toggling the tax rate,
- changing name,
- changing value.

jamesdeanexpensify replied:
I would guess, though, that we'd show the error message right after you take the action that triggered it, so it would be fairly easy to understand what went wrong. Maybe we can stick with the generics here: 

- `deleteFailureMessage`: An error occurred while deleting the tax rate. Please try again or ask Concierge for help.
- `updateFailureMessage`: An error occurred while updating the tax rate. Please try again or ask Concierge for help.
- `createFailureMessage`: An error occurred while creating the tax rate. Please try again or ask Concierge for help.

luacmartins replied:
Yea, agreed!

kosmydel replied:
I've updated the PR.

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+                value: {
+                    taxRates: {
+                        taxes: taxesIDsToUpdate.reduce<TaxRateEnabledMap>((acc, taxID) => {
+                            acc[taxID] = {isDisabled: !isEnabled, pendingFields: {isDisabled: null}, errorFields: {isDisabled: null}};
```
We don't need to update the `isDisabled` value since it was already done in `optimisticData`
```suggestion
                            acc[taxID] = {pendingFields: {isDisabled: null}, errorFields: {isDisabled: null}};
```

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+    | null
+>;
+
+/**
+ * API call to delete policy taxes
+ * @param taxesToDelete A tax IDs array to delete
+ */
```
NAB I don't think we need these docs since the function name is very descriptive and we have TS types for the params now

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+    const policyTaxRates = policy?.taxRates?.taxes;
+
+    if (!policyTaxRates) {
+        throw new Error('Policy or tax rates not found');
```
Let's just `console.debug` the error and return early

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+                    taxRates: {
+                        taxes: taxesToDelete.reduce<TaxRateDeleteMap>((acc, taxID) => {
+                            acc[taxID] = {
+                                pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE,
```
We should clear the `pendingAction` on delete case
```suggestion
                                pendingAction: null
```

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+                value: {
+                    taxRates: {
+                        taxes: {
+                            [taxID]: {value: stringTaxValue, pendingFields: {value: null}, errorFields: {value: null}},
```
Same here, no need to update the `value` field, we already did that in `optimisticData`
```suggestion
                            [taxID]: {pendingFields: {value: null}, errorFields: {value: null}},
```

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+    API.write(WRITE_COMMANDS.DELETE_POLICY_TAXES, parameters, onyxData);
+}
+
+/**
+ * Rename policy tax
+ */
```
```suggestion
```

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+        ],
+    };
+
+    if (!originalTaxRate.name) {
+        throw new Error('Tax rate name not found');
+    }
```
Why do we perform this check here?

kosmydel replied:
Before we called the API with the name, not the ID. I forgot to remove it.

luacmartins commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +182,299 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+                value: {
+                    taxRates: {
+                        taxes: {
+                            [taxID]: {name: newName, pendingFields: {name: null}, errorFields: {name: null}},
```
```suggestion
                            [taxID]: {pendingFields: {name: null}, errorFields: {name: null}},
```

luacmartins commented on src/pages/workspace/taxes/WorkspaceEditTaxPage.tsx:
```diff
@@ -0,0 +1,156 @@
+    const [isDeleteModalVisible, setIsDeleteModalVisible] = useState(false);
+    const canEdit = policy && canEditTaxRate(policy, taxID);
+
+    const toggle = () => {
```
```suggestion
    const toggleTaxRate = () => {
```

luacmartins commented on src/pages/workspace/taxes/WorkspaceEditTaxPage.tsx:
```diff
@@ -0,0 +1,156 @@
+        setPolicyTaxesEnabled(policyID, [taxID], !!currentTaxRate.isDisabled);
+    };
+
+    const deleteTax = () => {
```
```suggestion
    const deleteTaxRate = () => {
```

luacmartins commented on src/pages/workspace/taxes/WorkspaceTaxesPage.tsx:
```diff
@@ -66,34 +76,34 @@ function WorkspaceTaxesPage({policy, route}: WorkspaceTaxesPageProps) {
+                isSelected: !!selectedTaxesIDs.includes(key),
+                isDisabledCheckbox: !canEditTaxRate(policy, key),
+                isDisabled: value.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE,
+                pendingAction: value.pendingAction ?? (Object.keys(value.pendingFields ?? {}).length > 0 ? CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE : null),
```
Why do we have check for other pendingFields here? Isn't `value.pendingAction` enough?
```suggestion
                pendingAction: value.pendingAction ?? (Object.keys(value.pendingFields ?? {}).length > 0 ? CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE : null),
```

kosmydel replied:
When we update the tax, including toggling it or changing its name or value, we store this information in the `pendingFields` parameter, rather than in the `pendingAction`. This is done to indicate to the user exactly which field is pending an update.

It seems reasonable to assume that if a tax has a pending field update, we should also indicate this to the user on the tax list.

luacmartins commented on src/pages/workspace/taxes/WorkspaceTaxesPage.tsx:
```diff
@@ -182,13 +255,27 @@ function WorkspaceTaxesPage({policy, route}: WorkspaceTaxesPageProps) {
                         sections={[{data: taxesList, indexOffset: 0, isDisabled: false}]}
                         onCheckboxPress={toggleTax}
-                        onSelectRow={() => {}}
+                        onSelectRow={(tax: ListItem) => tax.keyForList && Navigation.navigate(ROUTES.WORKSPACE_TAXES_EDIT.getRoute(policyID, tax.keyForList))}
```
We should also deselect all taxes when navigating the user to the edit page, like we do for [categories](https://github.com/Expensify/App/blob/94a3111ea0369c1ff917e212f1b0d72cfa4f97fe/src/pages/workspace/categories/WorkspaceCategoriesPage.tsx#L141-L144)

luacmartins reviewed (changes requested):
Looking good. Just a few small comments.

kosmydel committed:
fix Spanish translations

kosmydel committed:
fix copies

DylanDylann commented:
@kosmydel I am reviewing. Please help to merge the latest main

kosmydel committed:
address review

kosmydel committed:
Merge branch 'main' into wave8/WorkspaceEditPage

kosmydel committed:
rename

DylanDylann commented on src/libs/actions/TaxRate.ts:
```diff
@@ -20,6 +31,41 @@ function covertTaxNameToID(name: string) {
+/**
+ * Whether the tax rate can be deleted and disabled
+ */
+function canEditTaxRate(policy: Policy, taxID: string): boolean {
```
if a function does not interact with ONYX, let's move it to PolicyUtils.ts. Also help to check other new function here

DylanDylann commented on src/pages/workspace/taxes/WorkspaceEditTaxPage.tsx:
```diff
@@ -0,0 +1,156 @@
+                            </View>
+                        </OfflineWithFeedback>
+                        <OfflineWithFeedback
+                            errors={ErrorUtils.getLatestErrorField(currentTaxRate, 'name')}
```
Why do we need to get latest error here?

kosmydel replied:
We want to display an error near the edited field which failed. I'm not sure I understand. Should we display the earliest error? The `src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx` has similar approach.

DylanDylann commented on src/libs/actions/TaxRate.ts:
```diff
@@ -39,7 +85,8 @@ function getNextTaxCode(name: string, taxRates?: TaxRates): string {
 
 function createPolicyTax(policyID: string, taxRate: TaxRate) {
     if (!taxRate.code) {
```
NAB: But I don't think we need this check

DylanDylann commented on src/libs/actions/TaxRate.ts:
```diff
@@ -119,10 +183,289 @@ function clearTaxRateError(policyID: string, taxID: string, pendingAction?: Pend
+    const policy = allPolicies?.[`${ONYXKEYS.COLLECTION.POLICY}${policyID}`];
+    const policyTaxRates = policy?.taxRates?.taxes;
+
+    if (!policyTaxRates) {
```
Same above

kosmydel replied:
I can add it, but it would result in optional chaining below:
```
taxNames: JSON.stringify(taxesToDelete.map((taxID) => policyTaxRates?.[taxID]?.name)),
```

DylanDylann commented on src/pages/workspace/taxes/ValuePage.tsx:
```diff
@@ -0,0 +1,96 @@
+                            defaultValue={value}
+                            onInputChange={setValue}
+                            hideCurrencySymbol
+                            extraSymbol={<Text style={styles.iouAmountText}>%</Text>}
```
```suggestion
                            extraSymbol={<Text style={styles.iouAmountText}>%</Text>}
                            autoFocus
```

kosmydel replied:
Adding it doesn't make the autofocus work. I have to investigate it.

DylanDylann commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

kosmydel committed:
rename taxes to tax

kosmydel committed:
move canEditTaxRate to PolicyUtils

kosmydel commented on src/types/form/WorkspaceTaxNameForm.ts:
```diff
@@ -0,0 +1,18 @@
+import type {ValueOf} from 'type-fest';
```
If needed, I can reuse it. Personally, I think, it might be confusing a bit to share it. cc @luacmartins

kosmydel commented:
I've addressed the review, there are two things left:
- more than two digits after the comma: https://github.com/Expensify/App/issues/38499#issuecomment-2004931877
- autofocus for the value page when editing.

kosmydel commented:
I've addressed the review, there are two things left:
- more than two digits after the comma: https://github.com/Expensify/App/issues/38499#issuecomment-2004931877
- autofocus for the value page when editing.

I will probably not have time today to work on this, as I am shorter today.

DylanDylann commented:
Will complete in one hour

DylanDylann commented:
@luacmartins LGTM!

kosmydel committed:
Merge branch 'main' into wave8/WorkspaceEditPage

kosmydel committed:
add FeatureEnabledAccessOrNotFoundWrapper

kosmydel commented:
Conflicts resolved, please test it one more time @DylanDylann, thanks!

luacmartins commented:
Tests were passing
![Screenshot 2024-03-20 at 9 17 15 AM](https://github.com/Expensify/App/assets/22219519/f0102c2a-9b7e-4aa9-b2d3-d83a491a368c)

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8393044290) to staging by https://github.com/luacmartins in version: 1.4.56-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8446744500) to production by https://github.com/yuwenmemon in version: 1.4.56-8 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

s77rt commented on src/ROUTES.ts:
```diff
@@ -616,6 +616,18 @@ const ROUTES = {
         route: 'settings/workspaces/:policyID/taxes/new',
         getRoute: (policyID: string) => `settings/workspaces/${policyID}/taxes/new` as const,
     },
+    WORKSPACE_TAX_EDIT: {
+        route: 'settings/workspaces/:policyID/tax/:taxID',
+        getRoute: (policyID: string, taxID: string) => `settings/workspaces/${policyID}/tax/${encodeURI(taxID)}` as const,
+    },
```
We should have used `encodeURIComponent` instead of `encodeURI` since `taxID` is not a uri. Also we should have added a decode step (similar to other logic e.g. tag).

`encodeURI` does not encode all special chars e.g. `/` which caused this bug https://github.com/Expensify/App/issues/39614

ahmedGaber93 commented on src/libs/PolicyUtils.ts:
```diff
@@ -277,6 +277,18 @@ function goBackFromInvalidPolicy() {
+ * Whether the tax rate can be deleted and disabled
+ */
+function canEditTaxRate(policy: Policy, taxID: string): boolean {
+    return policy.taxRates?.defaultExternalID !== taxID;
```
Coming from https://github.com/Expensify/App/issues/46657, here we prevent deleting `defaultExternalID` but missed to also prevent `foreignTaxDefault` and it fixed here https://github.com/Expensify/App/pull/46862

alitoshmatov commented on src/pages/workspace/taxes/NamePage.tsx:
```diff
@@ -0,0 +1,120 @@
+    const currentTaxRate = PolicyUtils.getTaxByID(policy, taxID);
+    const {inputCallbackRef} = useAutoFocusInput();
+
+    const [name, setName] = useState(() => parser.htmlToMarkdown(currentTaxRate?.name ?? ''));
```
We don't support markdown for tax names, removed this in https://github.com/Expensify/App/issues/54565
